### DISCUSSION TITLE
the motive is to integrate windows authentication for sql server using quarkus

### DISCUSSION THREAD

**Discussion by git4rputuval** [quarkusio/quarkus#38734] [Q&A]
Our quarkus version is 2.7.2.Final and mssql version is 4.2.4 (using the vertx-mssql-client-4.2.4 jar) but there is no proper documentation for how to let a user access the sql instance using windows authentication.

We did find a way for imperative programming but since we intend to do all of this with quarkus reactive we couldn't find any appropriate solution to this.

**Comment by sberyozkin** [comment]
@tsegismont Hi Thomas, do you have some recommendations ?

**Reply by tsegismont** [reply]
Thanks for the heads-up @sberyozkin

**Comment by tsegismont** [comment]
@git4rputuval are you talking about [NTLM authentication](https://learn.microsoft.com/en-us/sql/connect/jdbc/using-ntlm-authentication-to-connect-to-sql-server?view=sql-server-ver16)?

In this case, the Vert.x SQL Client for MSSQL does not support it yet. Feel free to file an [issue](https://github.com/eclipse-vertx/vertx-sql-client/), we may consider it, or I can guide you through the contribution process if you're interested.

**Reply by git4rputuval** [reply]
Yes, NTLM authentication applies for jdbc drivers only, and our entire application is built on top of Quarkus Reactive. Going back to JDBC based implementation is another 6 months project, and we have already completed UAT. Going back is not possible. Is there anyway we can get this working?

All our clients are undergoing security audits and pressing on windows authentication feature for App Server to DB connection. Sooner or later, all the projects implementing Quarkus Reactive is going to take a hit, if implementing in banking projects. 

I will create an issue.

**Reply by tsegismont** [reply]
I'm sorry to hear that, but it's not supported yet, and there is no workaround for this. The only supported auth mechanism is passwords.

**Reply by git4rputuval** [reply]
We have created the issue yesterday, [Windows Authentication Issue](https://github.com/eclipse-vertx/vertx-sql-client/issues/1407) for everybody's reference.

**Comment by git4rputuval** [comment]
> @git4rputuval are you talking about [NTLM authentication](https://learn.microsoft.com/en-us/sql/connect/jdbc/using-ntlm-authentication-to-connect-to-sql-server?view=sql-server-ver16)?
> 
> In this case, the Vert.x SQL Client for MSSQL does not support it yet. Feel free to file an [issue](https://github.com/eclipse-vertx/vertx-sql-client/), we may consider it, or I can guide you through the contribution process if you're interested.

Is there any implementation like the credential provider (https://quarkus.io/guides/credentials-provider) that we can use to get access to the connection pool and intercept it before the pool request is sent from the application server to database?

**Reply by tsegismont** [reply]
With the credentials provider, you can implement rotating passwords, which might be acceptable by your clients for their security audits.

**Reply by git4rputuval** [reply]
This is for a secured application start up. If we implement the Windows Authentication on the Application Server/ Undertow, only the authorized user can login into the server machine and start or stop the application server.

The application server may be started by any user. But since the database connection pool could not be started with the logged in user, the application server is rendered ineffective. The reactive application will not be able to query or update the database since the connection pool could not be started because the logged in user is not a windows-authenticated SQL user whose credentials will be used to create the connection pool.

**Comment by git4rputuval** [comment]
We are attaching a demo project which demonstrates our application architecture to a small extent to reproduce the errors we're getting. This project attachment is giving a runtime error whereas in our actual application its a compilation error, we are attaching both errors as snapshots, we expect that if either of the error is fixed the other would also be solved.

Below is the demo application
[persistenceUnit_Error.zip](https://github.com/eclipse-vertx/vertx-sql-client/files/14402705/persistenceUnit_Error.zip)

the compilation error in our actual application is 
<img width="960" alt="compilationErrorForSimulatenousDataSources" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/6c3c4348-b4f0-4a46-b989-a9c11d1381fe">

the runtime error wrt this demo application is
<img width="960" alt="runtimeErrorWithSimlutaneousDataSources" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/3473fca7-d385-4609-8124-20bc13535b42">

this is runtime error when we disable reactive datasource using quarkus.datasource.reactive=false in our application.properties file
<img width="725" alt="disableReactiveDataSource" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/cb72f0e4-8602-4708-99ac-0fdbf1e9d355">

<img width="960" alt="runtimeErrorWith ReactiveDisabled" src="https://github.com/eclipse-vertx/vertx-sql-client/assets/96055225/fce78253-f4b3-423b-9910-06d691b9a4c2">

The solutions we used by far involved using @PersistenceUnit to attach our entity/model to a specific data source, made two different entities one to map if persistence is done using jdbc and depending on the flag value set in application.properties persistence through our data service layer would either happen from jdbc persistence side (jdbcEntity,jdbcService,jdbcRepo classes) else the usual/preffered reactive side.
